Generate cluster details in JSON output

parthosa opened the pull request:
Fixes #911  👮‍♂️. This PR adds additional cluster information fields in the `cluster_information.json` file:

**Fields Added:**
1. `driverHost`
2. `clusterId`
3. `clusterName`

Note: Depending on platform, some of these fields may be unavailable.

## Platform Based Availability
 
 
| Platform         | driverHost | clusterId | clusterName |
|------------------|------------|-----------|-------------|
| Databricks AWS   | Yes        | Yes       | Yes         |
| Databricks Azure | Yes        | Yes       | Yes         |
| EMR              | Yes        | Yes       |    :x:         |
| Dataproc         | Yes        |   :x:        |      :x:       |
| OnPrem           | Yes        |     :x:      |    :x:         |


## Test

### Manual Testing

#### 1. Databricks AWS

```
[ {
  "appName" : "Databricks Shell",
  "appId" : "app-20231212214826-0000",
  "eventLogPath" : "file:/Users/username/Work/event-logs/databricks-aws-cpu",
  "clusterInfo" : {
    "coresPerExecutor" : 8,
    "numExecutorNodes" : 2,
    "executorInstance" : "m6gd.2xlarge",
    "driverInstance" : "m6gd.xlarge",
    "driverHost" : "ip0.ip1.ip2.ip3",
    "clusterId" : "1212-214324-MY_CLUSTER_ID",
    "clusterName" : "MY_CLUSTER_NAME"
  }
} ]  
```

#### 2. EMR
```
[ {
  "appName" : "NDS - Power Run",
  "appId" : "application_1692643187882_0001",
  "eventLogPath" : "file:/Users/username/Work/event-logs/emr-cpu/application_1692643187882_0001",
  "clusterInfo" : {
    "coresPerExecutor" : 16,
    "numExecutorNodes" : 8,
    "driverHost" : "MY_DRIVER_HOST",
    "clusterId" : "MY_CLUSTER_ID"
  }
} ]
```

#### 3. Dataproc:

```
[ {
  "appName" : "NDS - query72",
  "appId" : "application_1686676198636_0002",
  "eventLogPath" : "file:/Users/username/Work/event-logs/dataproc-cpu/eventlog-test-1",
  "clusterInfo" : {
    "coresPerExecutor" : 16,
    "numExecutorNodes" : 4,
    "driverHost" : "MY_DRIVER_HOST"
  }
} ]
```

### Unit Tests

Unit tests were updated to include platform based cluster information.

cindyyuanjiang commented:
How are the test event logs generated?

amahussein reviewed (approved):
Thanks @parthosa !

amahussein reviewed (changes requested):
Thanks @parthosa 
It will be nice to have an entry to indicate the platform.
For example, something like "`vendor`" or `provider` to represent the CSP platform on which the eventlog was generated.
This will make it easy for the consumer of the JSON file to identify the platform without need to do further mapping.

As a bonus: It will be nice to make this information part of the Application itself.
If it becomes an entry in the App properties, then
- SQLPlanParser can use it to make decisions on some operators support decisions. Furthermore, we will be able to customize the support per vendor when needed.
- the AuttoTuner can easily consume it to make some of the calculations. I understand that we pass the platform as an argument to the AutoTuner, but we might need that feature later.

If this bonus requires a big change, then I am OK with this PR given that we add a field to indicate the vendor.

Let me know if you have any question.

parthosa commented:
> How are the test event logs generated?

I used existing event logs and added these properties in them. Ideally I should have used actual event logs for each platform and removed the unnecessary parts.

cindyyuanjiang reviewed (approved):
Thanks @parthosa! LGTM

parthosa commented on core/src/main/scala/org/apache/spark/sql/rapids/tool/qualification/QualificationAppInfo.scala:
```diff
@@ -62,8 +62,16 @@ class QualificationAppInfo(
   var clusterTags: String = ""
   var clusterTagClusterId: String = ""
   var clusterTagClusterName: String = ""
+  var clusterInfo: Option[ClusterInfo] = None
```
Currently cluster detection is implemented only for Q tool. 

We can move to `AppBase` with some refactoring. `Platform` must be made available to `AppBase`

amahussein replied:
No worries.
I believe this refactor could be part of the Tuning revamp since we will need the HW configurations to suggest different HW shape for tuning.
I will work on that next Spring.

parthosa replied:
Thanks @amahussein.
